File: clue/deduction.cpp
#include "deduction.h"
#include <iostream>
#include <vector>
#include <string>

using namespace std;

deduction::deduction(vector<string> names)
    : f(names)
{
}

void deduction::deduce(vector<pair<vector<vector<string>>, pair<bool, string>>>& accusations,
    vector<string> names, vector<vector<string>>& cards)
{
    // have to add facts about cards you already know/have
    for (auto card : cards)
    {
        vector<pair<string, string>> temp;
        temp.push_back({ card[0], card[1] });
        f.addFact(temp);
    }

    // each accusation needs to add some constraints to hashmap
    for (auto accusation : accusations)
    {
        // gather the info
        vector<vector<string>> info = accusation.first;
        vector<string> moreInfo = info[0];
        string accuser = moreInfo[0];
        string person = moreInfo[1];
        string weapon = moreInfo[2];
        string place = moreInfo[3];
        vector<string> playersInvolved = info[1];
        bool showedCard = accusation.second.first;
        string knownCard = accusation.second.second;

	    // add fact about players who had something
        int numPlayersInvolved = playersInvolved.size();
        if (showedCard)
        {
            string playerWithSomething = playersInvolved[numPlayersInvolved - 1];
            vector<pair<string, string>> temp1;

            if (accuser == names[0])
            {
                temp1.push_back({ playerWithSomething, knownCard });
            }
            else
            {
                temp1.push_back({ playerWithSomething, person });
                temp1.push_back({ playerWithSomething, weapon });
                temp1.push_back({ playerWithSomething, place });
            }
            
            f.addFact(temp1);
        }

        // add facts about players with nothing
        int iterations = numPlayersInvolved - 1;
        bool addEnvelope = false;
        if (!showedCard)
        {
            ++iterations;
            addEnvelope = true;
        }

        for (int i = 0; i < iterations; ++i)
        {
            string currPerson = playersInvolved[i];
            vector<pair<string, string>> tempPerson;
            vector<pair<string, string>> tempWeapon;
            vector<pair<string, string>> tempPlace;

            // if nobody showed the card, the only people that could have it is the accuser or the envelope
            if (addEnvelope)
            {
                // person
                tempPerson.push_back({ "envelope", person });
                tempPerson.push_back({ accuser, person });

                // weapon
                tempWeapon.push_back({ "envelope", weapon });
                tempWeapon.push_back({ accuser, weapon });

                // place
                tempPlace.push_back({ "envelope", place });
                tempPlace.push_back({ accuser, place });
            }
            else
            {
                for (string player : names)
                {
                    // add a fact involving the current card and all people that is not the current person
                    if (player != currPerson)
                    {
                        // person
                        tempPerson.push_back({ player, person });

                        // weapon
                        tempWeapon.push_back({ player, weapon });

                        // place
                        tempPlace.push_back({ player, place });
                    }
                }

                // account for envelope having the cards as well
                tempPerson.push_back({ "envelope", person });
                tempWeapon.push_back({ "envelope", weapon });
                tempPlace.push_back({ "envelope", place });
            }

            f.addFact(tempPerson);
            f.addFact(tempWeapon);
            f.addFact(tempPlace);
        }
    }

    // look through facts and do some deductions
    f.eliminate(cards, names);
}

File: web-clue/game/Deduction.js
const Facts = require('./Facts');

class Deduction {
    constructor() {
        this.facts = null;
    }
    
    deduce(accusations, playerNames, cards) {
        console.log('=== DEDUCTION DEBUG ===');
        console.log('accusations:', JSON.stringify(accusations, null, 2));
        console.log('playerNames:', playerNames);
        console.log('cards:', cards);
        
        // Initialize facts with player names
        this.facts = new Facts(playerNames);
        
        // Add facts about cards we already know/have
        for (const card of cards) {
            const temp = [[card[0], card[1]]];
            this.facts.addFact(temp);
        }
        
        // Process each accusation to add constraints
        for (const accusation of accusations) {
            console.log('Processing accusation:', accusation);
            this.processAccusation(accusation, playerNames);
        }
        
        // Perform elimination and return new deductions
        const newDeductions = this.facts.eliminate(cards, playerNames);
        console.log('newDeductions:', newDeductions);
        console.log('=== END DEDUCTION DEBUG ===');
        return newDeductions;
    }
    
    processAccusation(accusation, playerNames) {
        // Extract accusation info
        const accusationDetails = accusation.accusationDetails[0];
        const accuser = accusationDetails[0];
        const suspect = accusationDetails[1];
        const weapon = accusationDetails[2];
        const room = accusationDetails[3];
        const playersInvolved = accusation.accusationDetails[1];
        const cardShown = accusation.result.cardShown;
        const knownCard = accusation.result.knownCard;
        
        const numPlayersInvolved = playersInvolved.length;
        
        // Add fact about players who had something
        if (cardShown) {
            const playerWithCard = playersInvolved[numPlayersInvolved - 1];
            let possibilities = [];
            
            if (accuser === playerNames[0]) {
                // If we are the accuser, we know exactly which card was shown
                possibilities.push([playerWithCard, knownCard]);
            } else {
                // We don't know which specific card, but we know they have one of these
                possibilities.push([playerWithCard, suspect]);
                possibilities.push([playerWithCard, weapon]);
                possibilities.push([playerWithCard, room]);
            }
            
            this.facts.addFact(possibilities);
        }
        
        // Add facts about players with nothing
        let iterations = numPlayersInvolved - 1;
        let addEnvelope = false;
        
        if (!cardShown) {
            iterations++;
            addEnvelope = true;
        }
        
        for (let i = 0; i < iterations; i++) {
            const currentPlayer = playersInvolved[i];
            
            if (addEnvelope) {
                // Nobody showed the card, so only the accuser or envelope could have it
                const suspectPossibilities = [
                    ["envelope", suspect],
                    [accuser, suspect]
                ];
                const weaponPossibilities = [
                    ["envelope", weapon],
                    [accuser, weapon]
                ];
                const roomPossibilities = [
                    ["envelope", room],
                    [accuser, room]
                ];
                
                this.facts.addFact(suspectPossibilities);
                this.facts.addFact(weaponPossibilities);
                this.facts.addFact(roomPossibilities);
            } else {
                // This player doesn't have any of the cards, so other players or envelope must have them
                const suspectPossibilities = [];
                const weaponPossibilities = [];
                const roomPossibilities = [];
                
                for (const player of playerNames) {
                    if (player !== currentPlayer) {
                        suspectPossibilities.push([player, suspect]);
                        weaponPossibilities.push([player, weapon]);
                        roomPossibilities.push([player, room]);
                    }
                }
                
                // Account for envelope having the cards as well
                suspectPossibilities.push(["envelope", suspect]);
                weaponPossibilities.push(["envelope", weapon]);
                roomPossibilities.push(["envelope", room]);
                
                this.facts.addFact(suspectPossibilities);
                this.facts.addFact(weaponPossibilities);
                this.facts.addFact(roomPossibilities);
            }
        }
    }
}

module.exports = Deduction;